From 85ac579dc40b3ede34a9178f7d7533468f51e309 Mon Sep 17 00:00:00 2001
From: Liu Yibin <yibin_liu@c-sky.com>
Date: Mon, 28 Oct 2019 10:33:30 +0800
Subject: [PATCH] Ebreak in poweroff

Signed-off-by: Liu Yibin <yibin_liu@c-sky.com>
---
 machine/mtrap.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/machine/mtrap.c b/machine/mtrap.c
index 4ba1ec9..2e688ee 100644
--- a/machine/mtrap.c
+++ b/machine/mtrap.c
@@ -345,6 +345,7 @@ void trap_from_machine_mode(uintptr_t* regs, uintptr_t dummy, uintptr_t mepc)
 
 void poweroff(uint16_t code)
 {
+  asm volatile ("ebreak");
   printm("Power off\r\n");
   finisher_exit(code);
   if (htif) {
-- 
2.7.4

