From a012e4192439c619c79af97d5b1bdbeb74f4991f Mon Sep 17 00:00:00 2001
From: Guo Ren <guoren@linux.alibaba.com>
Date: Thu, 9 Jan 2020 14:36:46 +0800
Subject: [PATCH 04/10] riscv: Support T-HEAD XuanTie CPU sbi smp boot

Signed-off-by: Guo Ren <guoren@linux.alibaba.com>
---
 arch/riscv/kernel/smpboot.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/arch/riscv/kernel/smpboot.c b/arch/riscv/kernel/smpboot.c
index 261f408..a1b09be 100644
--- a/arch/riscv/kernel/smpboot.c
+++ b/arch/riscv/kernel/smpboot.c
@@ -102,6 +102,7 @@ int __cpu_up(unsigned int cpu, struct task_struct *tidle)
 	int hartid = cpuid_to_hartid_map(cpu);
 	tidle->thread_info.cpu = cpu;
 
+	SBI_CALL_1(0x09000003, hartid);
 	/*
 	 * On RISC-V systems, all harts boot on their own accord.  Our _start
 	 * selects the first hart to boot the kernel and causes the remainder
-- 
2.7.4

