From 06859196e3156b293c67b4fd465e8541497af428 Mon Sep 17 00:00:00 2001
From: Guo Ren <ren_guo@c-sky.com>
Date: Sat, 6 Jul 2019 23:56:25 +0800
Subject: [PATCH 01/15] package/Config.in: add Config.in.csky

Signed-off-by: Guo Ren <ren_guo@c-sky.com>
Signed-off-by: Guo Ren <guoren@linux.alibaba.com>
---
 package/Config.in | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/package/Config.in b/package/Config.in
index 4ddde0d..6dea045 100644
--- a/package/Config.in
+++ b/package/Config.in
@@ -1,3 +1,5 @@
+	source "package/Config.in.csky"
+
 menu "Target packages"
 
 	source "package/busybox/Config.in"
-- 
2.17.1

