From 0025142fbb2cf18a3d120c698ff6260b05286672 Mon Sep 17 00:00:00 2001
From: Guo Ren <ren_guo@c-sky.com>
Date: Sat, 28 Dec 2019 18:53:25 +0800
Subject: [PATCH] linux/linux.mk: Copy Image to images

Signed-off-by: Guo Ren <ren_guo@c-sky.com>
---
 linux/linux.mk | 1 +
 1 file changed, 1 insertion(+)

diff --git a/linux/linux.mk b/linux/linux.mk
index a137834..77463fa 100644
--- a/linux/linux.mk
+++ b/linux/linux.mk
@@ -468,6 +468,7 @@ else
 # build process.
 define LINUX_INSTALL_IMAGE
 	$(INSTALL) -m 0644 -D $(LINUX_IMAGE_PATH) $(1)/$(LINUX_IMAGE_NAME)
+	$(INSTALL) -m 0644 -D $(LINUX_ARCH_PATH)/boot/Image $(1)/Image
 endef
 endif
 
-- 
2.17.0

