From 207e145d3bc16ce2b2bda75430ac08519314ef8e Mon Sep 17 00:00:00 2001
From: Guo Ren <guoren@linux.alibaba.com>
Date: Thu, 21 Jan 2021 10:52:33 +0800
Subject: [PATCH 16/16] boot/opensbi: Down version to 0.6

Signed-off-by: Guo Ren <guoren@linux.alibaba.com>
---
 boot/opensbi/opensbi.mk | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/boot/opensbi/opensbi.mk b/boot/opensbi/opensbi.mk
index 6af7ebd..ed79d48 100644
--- a/boot/opensbi/opensbi.mk
+++ b/boot/opensbi/opensbi.mk
@@ -4,7 +4,7 @@
 #
 ################################################################################
 
-OPENSBI_VERSION = 0.8
+OPENSBI_VERSION = 0.6
 OPENSBI_SITE = $(call github,riscv,opensbi,v$(OPENSBI_VERSION))
 OPENSBI_LICENSE = BSD-2-Clause
 OPENSBI_LICENSE_FILES = COPYING.BSD
-- 
2.17.1

