From d800d1157812f222230bd4687832992c39d700dd Mon Sep 17 00:00:00 2001
From: Liu Yibin <yibin_liu@c-sky.com>
Date: Thu, 18 Apr 2019 09:53:01 +0800
Subject: [PATCH 1/4] mapping 1GB DDR

Signed-off-by: Liu Yibin <yibin_liu@c-sky.com>
---
 arch/riscv/Kconfig | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/arch/riscv/Kconfig b/arch/riscv/Kconfig
index eb56c82..eca0657 100644
--- a/arch/riscv/Kconfig
+++ b/arch/riscv/Kconfig
@@ -61,6 +61,7 @@ config PAGE_OFFSET
 	hex
 	default 0xC0000000 if 32BIT && MAXPHYSMEM_2GB
 	default 0xffffffff80000000 if 64BIT && MAXPHYSMEM_2GB
+	default 0xffffffffc0000000 if 64BIT && MAXPHYSMEM_1GB
 	default 0xffffffe000000000 if 64BIT && MAXPHYSMEM_128GB
 
 config STACKTRACE_SUPPORT
@@ -151,6 +152,8 @@ choice
 	default MAXPHYSMEM_2GB if 64BIT && CMODEL_MEDLOW
 	default MAXPHYSMEM_128GB if 64BIT && CMODEL_MEDANY
 
+	config MAXPHYSMEM_1GB
+		bool "1GiB"
 	config MAXPHYSMEM_2GB
 		bool "2GiB"
 	config MAXPHYSMEM_128GB
-- 
2.7.4

