From 43f2a6c0a236fbc7f8d3fbb3e25c40a7b8d0c4df Mon Sep 17 00:00:00 2001
From: Liu Yibin <yibin_liu@c-sky.com>
Date: Fri, 3 Jan 2020 16:44:13 +0800
Subject: [PATCH] lib: Ignore Vendor sbi call

Signed-off-by: Liu Yibin <yibin_liu@c-sky.com>
---
 lib/sbi_ecall.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/lib/sbi_ecall.c b/lib/sbi_ecall.c
index 42123ba..d79af98 100644
--- a/lib/sbi_ecall.c
+++ b/lib/sbi_ecall.c
@@ -75,6 +75,8 @@ int sbi_ecall_handler(u32 hartid, ulong mcause,
 		ret = 0;
 		break;
 	default:
+		if ((regs->a7 & 0xff000000) == 0x09000000)
+			ret = 0;
 		break;
 	};
 
-- 
2.17.0

