From 528bf0275e6fc65d14bbe74d9cd8a78d2aec1c10 Mon Sep 17 00:00:00 2001
From: Ma Jun <majun258@linux.alibaba.com>
Date: Tue, 10 Mar 2020 19:44:50 +0800
Subject: [PATCH] Add the csky arch support

Signed-off-by: Ma Jun <majun258@linux.alibaba.com>
---
 include/private/gcconfig.h | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/include/private/gcconfig.h b/include/private/gcconfig.h
index 1fdf3d3..9fa2be8 100644
--- a/include/private/gcconfig.h
+++ b/include/private/gcconfig.h
@@ -662,6 +662,11 @@ EXTERN_C_BEGIN
 #   define mach_type_known
 # endif
 
+# if defined(__csky__) && defined(LINUX)
+#   define CSKY
+#   define mach_type_known
+# endif
+
 # if defined(SN_TARGET_PSP2)
 #   define mach_type_known
 # endif
@@ -2868,6 +2873,19 @@ EXTERN_C_BEGIN
 #   endif
 # endif /* RISCV */
 
+# ifdef CSKY
+#   define MACH_TYPE "CSKY"
+#   define CPP_WORDSZ 32  /* 32 or 64 */
+#   define ALIGNMENT (CPP_WORDSZ/8)
+#   ifdef LINUX
+#     define OS_TYPE "LINUX"
+      extern int __data_start[];
+#     define DATASTART ((ptr_t)__data_start)
+#     define LINUX_STACKBOTTOM
+#     define DYNAMIC_LOADING
+#   endif
+# endif /* CSKY */
+
 #if defined(__GLIBC__) && !defined(DONT_USE_LIBC_PRIVATES)
   /* Use glibc's stack-end marker. */
 # define USE_LIBC_PRIVATES
-- 
1.8.3.1

